src/main/java/uk/gov/ch/developer/docs/controller/developer/SignInController.java

```diff
@@ -0,0 +1,231 @@
+import uk.gov.companieshouse.logging.Logger;
+import uk.gov.companieshouse.logging.LoggerFactory;
+import uk.gov.companieshouse.session.Session;
+import uk.gov.companieshouse.session.SessionImpl;
+import uk.gov.companieshouse.session.SessionKeys;
+import uk.gov.companieshouse.session.handler.SessionHandler;
+import uk.gov.companieshouse.session.model.SignInInfo;
+import uk.gov.companieshouse.session.model.UserProfile;
+
+@Controller
+@RequestMapping("/signin")
+public class SignInController {
+
+    private static final String OAUTH_COMPANY_SCOPE_PREFIX =
+            "https://api.companieshouse.gov.uk/company/";
+
+    private static final EnvironmentReader reader = new EnvironmentReaderImpl();
+
+    private static final String base64Key = reader.getMandatoryString("OAUTH2_REQUEST_KEY");
+    private static final String authorizationUri = reader.getMandatoryString("OAUTH2_AUTH_URI");
+    private static final String clientId = reader.getMandatoryString("OAUTH2_CLIENT_ID");
+    private static final String redirectUri = reader.getMandatoryString("OAUTH2_REDIRECT_URI");
+    private static final String secret = reader.getMandatoryString("COOKIE_SECRET");
+
+    protected Logger LOGGER = LoggerFactory.getLogger("docs.developer.ch.gov.uk");
+
+    @GetMapping
+    public void getSignIn(ServletRequest request, ServletResponse response) throws IOException {
+
+        Session chSession =
+                (Session) request.getAttribute(SessionHandler.CHS_SESSION_REQUEST_ATT_KEY);
+
+        HttpServletRequest httpServletRequest = (HttpServletRequest) request;
+        HttpServletResponse httpServletResponse = (HttpServletResponse) response;
+
+        // Redirect for user authentication (no scope specified)
+        redirectForAuth(chSession, httpServletRequest, httpServletResponse, null, false);
+
+    }
+
+    /**
+     * Redirects to a URI for the user to authenticate themselves
+     * 
+     * @param session The user's session, retrieved from context
+     * @param request
+     * @param response
+     * @param companyNumber The company number a user is viewing
+     * @param force Boolean determining whether to force re-authentication
+     * @throws IOException
+     */
+    protected void redirectForAuth(Session session, HttpServletRequest request,
+            HttpServletResponse response, String companyNumber, boolean force) throws IOException {
+
+        // Find the original requested url
+        StringBuilder originalRequestUrl = new StringBuilder(request.getRequestURL().toString());
+        String queryString = request.getQueryString();
+        if (queryString != null)
+            originalRequestUrl.append("?").append(queryString);
+
+        // Set the scope
+        String scope = null;
+        if (companyNumber != null)
+            scope = OAUTH_COMPANY_SCOPE_PREFIX + companyNumber;
+
+        // Generate and store a nonce in the session
+        Session sessionToUpdate = session;
```

Comment:
Why a new variable? Why do we get username from the older one? Can't we just use the one that was passed in?

Reply:
This is carrying existing functionality from the filters, not quite sure why it was done this way. Perhaps an extra layer of security, protecting against possible hijacking? As the nonce will be stored in this new session and not the one retrieved from context.